lib/core/src/Cardano/Wallet/Primitive/Types/TokenPolicy.hs

```diff
@@ -0,0 +1,96 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE DerivingVia #-}
+
+module Cardano.Wallet.Primitive.Types.TokenPolicy
+    (
+      -- * Token Policies
+      TokenPolicyId
+    , mkTokenPolicyId
+
+      -- * Token Names
+    , TokenName
+    , mkTokenName
+
+    ) where
+
+import Prelude
+
+import Cardano.Wallet.Primitive.Types.Hash
+    ( Hash (..) )
+import Control.DeepSeq
+    ( NFData )
+import Control.Monad
+    ( (>=>) )
+import Data.Aeson
+    ( FromJSON (..), ToJSON (..) )
+import Data.ByteString
+    ( ByteString )
+import Data.Text
+    ( Text )
+import Data.Text.Class
+    ( FromText (..), ToText (..) )
+import Fmt
+    ( Buildable (..) )
+import GHC.Generics
+    ( Generic )
+import Quiet
+    ( Quiet (..) )
+
+--------------------------------------------------------------------------------
+-- Token policy identifiers
+--------------------------------------------------------------------------------
+
+newtype TokenPolicyId = TokenPolicyId
+    { unTokenPolicyId :: Hash "TokenPolicy" }
+    deriving stock (Eq, Ord, Generic)
+    deriving (Read, Show) via (Quiet TokenPolicyId)
+
+instance NFData TokenPolicyId
+
+instance Buildable TokenPolicyId where
+    build = build . toText . unTokenPolicyId
+
+instance FromJSON TokenPolicyId where
+    parseJSON = parseJSON >=> either (fail . show) pure . fromText
+
+instance ToJSON TokenPolicyId where
+    toJSON = toJSON . toText
+
+instance ToText TokenPolicyId where
+    toText = toText . unTokenPolicyId
+
+instance FromText TokenPolicyId where
+    fromText = fmap TokenPolicyId . fromText
+
+mkTokenPolicyId :: ByteString -> TokenPolicyId
+mkTokenPolicyId = TokenPolicyId . Hash
```

Comment:
> This is arguably the unsafe way to do this. I'd expect the safe constructor to be TokenPolicy -> TokenPolicyId

@KtorZ Thanks for spotting this. I had written this function very early on, and intended to make it safer later on.

I've now removed the function entirely.

Since we don't yet have a `TokenPolicy` type, I'll leave the creation of a `TokenPolicy -> TokenPolicyId` function to a future PR.

Fixed in f7d596eb3ab4920b4b9edb2c6aff4797be8b9182.